epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyVisibilityTracker.java

```diff
@@ -155,22 +241,88 @@ public void onLayoutChange(
         int left, int top, int right, int bottom,
         int oldLeft, int oldTop, int oldRight, int oldBottom
     ) {
-      processChildren();
+      processChangeEvent("onLayoutChange");
     }
 
     @Override
     public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
-      processChildren();
+      processChangeEvent("onScrolled");
     }
 
     @Override
     public void onChildViewAttachedToWindow(View child) {
-      processChild(child);
+      processChangeEvent("onChildViewAttachedToWindow");
     }
 
     @Override
     public void onChildViewDetachedFromWindow(View child) {
-      processChild(child, true);
+      // On detach event send the detached view
+      processChangeEventWithDetachedView(child, "onChildViewDetachedFromWindow");
+    }
+  }
+
+  /**
+   * The layout/scroll events are not enough to detect all sort of visibility changes. We also
+   * need to look at the data events from the adapter.
+   */
+  class DataObserver extends AdapterDataObserver {
+
+    /**
+     * Clear the current visibility statues
+     */
+    @Override
+    public void onChanged() {
+      if (DEBUG_LOG) {
+        Log.d(TAG, "onChanged()");
+      }
+      visibilityIdToItemMap.clear();
+      visibilityIdToItems.clear();
+    }
+
+    /**
+     * For all items after the inserted range shift each {@link EpoxyVisibilityTracker} adapter
+     * position by inserted item count.
+     */
+    @Override
+    public void onItemRangeInserted(int positionStart, int itemCount) {
+      if (DEBUG_LOG) {
+        Log.d(TAG, String.format("onItemRangeInserted(%d, %d)", positionStart, itemCount));
+      }
+      for (EpoxyVisibilityItem item : visibilityIdToItems) {
+        if (item.getAdapterPosition() >= positionStart) {
+          item.shiftBy(itemCount);
+        }
+      }
+    }
+
+    /**
+     * For all items after the removed range reverse-shift each {@link EpoxyVisibilityTracker}
+     * adapter position by removed item count
+     */
+    @Override
+    public void onItemRangeRemoved(int positionStart, int itemCount) {
+      if (DEBUG_LOG) {
+        Log.d(TAG, String.format("onItemRangeRemoved(%d, %d)", positionStart, itemCount));
+      }
+      for (EpoxyVisibilityItem item : visibilityIdToItems) {
+        if (item.getAdapterPosition() >= positionStart) {
+          item.shiftBy(-itemCount);
+        }
+      }
+    }
+
+    @Override
+    public void onItemRangeMoved(int fromPosition, int toPosition, int itemCount) {
+      if (DEBUG_LOG) {
+        Log.d(TAG,
+            String.format("onItemRangeMoved(%d, %d, %d)", fromPosition, toPosition, itemCount));
+      }
+      for (EpoxyVisibilityItem item : visibilityIdToItems) {
+        int position = item.getAdapterPosition();
```

Comment:
other items can be affected too. moving an item from position 0 to 10 shifts items 1-9 up by one